### ISSUE TITLE
Multi-step forecasting horizons with multiple cutoffs/window

### ISSUE THREAD

**Issue by ngupta23** [alan-turing-institute/sktime#633]
**Is your feature request related to a problem? Please describe.**
I was wondering if there is a roadmap for addition of multi-step forecasting horizons with multiple cutoffs/window. This will be very helpful for `ForecastingGridSearchCV`

https://github.com/alan-turing-institute/sktime/blob/79ac80ab65adb30fdf233e98433646c1989da2fc/sktime/forecasting/model_selection/_tune.py#L37

**Current**
```python
y = load_airline()

fh = np.arange(1, 13)
y_train, y_test = temporal_train_test_split(y, test_size=len(fh))
print(y.shape, y_train.shape[0], y_test.shape[0])

cv = SlidingWindowSplitter(initial_window=int(len(y_train) * 0.5), start_with_window=True)

regressor = RandomForestRegressor(random_state=42)
forecaster = ReducedRegressionForecaster(regressor=regressor, strategy="recursive")

param_grid = {"regressor__n_estimators": [200, 300]}

gscv = ForecastingGridSearchCV(forecaster=forecaster, cv=cv, param_grid=param_grid, verbose=1)
gscv.fit(y_train)
```

The current flow only seems to be able to perform "one step ahead" validation

![image](https://user-images.githubusercontent.com/33585645/105845615-aefaa500-5fa0-11eb-9d5a-3e0e3ae90c4c.png)

**Describe the solution you'd like**

```python
y = load_airline()

fh = np.arange(1, 13)  # How much to forecast (from 1 to 12 or 1 year)
y_train, y_test = temporal_train_test_split(y, test_size=len(fh))
print(y.shape, y_train.shape[0], y_test.shape[0])

# Want to provide the forecast horizon here
cv = SlidingWindowSplitter(initial_window=int(len(y_train) * 0.5), fh=fh, start_with_window=True)

regressor = RandomForestRegressor(random_state=42)
forecaster = ReducedRegressionForecaster(regressor=regressor, strategy="recursive")

param_grid = {"regressor__n_estimators": [200, 300]}

gscv = ForecastingGridSearchCV(forecaster=forecaster, cv=cv, param_grid=param_grid, verbose=1)
gscv.fit(y_train)
```

Ideally, the validation should be done for the entire horizon in each window. 

![image](https://user-images.githubusercontent.com/33585645/105846155-6099d600-5fa1-11eb-9b91-066f61f8dfcb.png)

**Describe alternatives you've considered**
None that I can think of

**Comment by mloning** [comment]
Hi @ngupta23, the main problem was the representation of the CV results and aggregation of performance metrics. How would you want the CV results to look like?

**Comment by ngupta23** [comment]
Hi @mloning . Thanks for getting back. The basic idea is to have a reference score across a realistic forecast window and multiple folds so that multiple models can be compared and stability of the algorithm can also be tested. So you could simply return the list of metrics across the various folds along with some summary numbers like mean, median and standard deviation (to determine stability of algorithm across various folds). 

Using this information, the user (or the library itself) could provide some utilities to plot this such as the examples [shown here](https://topepo.github.io/caret/model-training-and-tuning.html#model-training-and-parameter-tuning) (albeit they are for classification)

### Plot 1
![image](https://user-images.githubusercontent.com/33585645/105907899-248b6300-5feb-11eb-96dd-655857992106.png)

### Plot 2
![image](https://user-images.githubusercontent.com/33585645/105907983-4553b880-5feb-11eb-8e90-47a3f2312e88.png)

### Raw Data Summary Stats
![image](https://user-images.githubusercontent.com/33585645/105908092-6e744900-5feb-11eb-9d25-368477ca7d6c.png)

**Comment by fkiraly** [comment]
@ngupta23, I am having trouble understanding what you are proposing.

First, are you proposing:
* a tuning method (compositor)
* an evaluation workflow?

Second, the devil is in the detail when you talk about error estimation using re-sample estimates - and it's easy to get confused too.

Would you hence mind to explain precisely the object you are talking about (e.g., proposing), optimally using mathematical formalism? Alternatively, pseudocode, but math allows you to be more precise typically.

**Comment by ngupta23** [comment]
Hi @fkiraly 

> First, are you proposing:

> a tuning method (compositor)
> an evaluation workflow?

If I understand your question correctly, I am not proposing anything different from what is already there in the flow today. Based on my understanding, the workflow for `ForecastingGridSearchCV` is doing both today - running the model for multiple hyperparameters (tuning) and then selecting the best set of hyperparameters based on the evaluation in the validation period. 

All I am asking is if the forecast horizon can be extended to support a realistic horizon based on end user's forecast horizon instead of a "one step ahead" (or a single point) validation. If I try to put in a forecast horizon of 12 periods in the code, it gives an error that this is not supported yet.



> Second, the devil is in the detail when you talk about error estimation using re-sample estimates - and it's easy to get confused too.

Maybe the images complicated the discussion :). So let me rephrase 

Lets say my forecast horizon is 12 periods (1 year with monthly data). So for fold 1, I will get 12 forecast values. This can be compared with the actual values for these 12 periods and sMAPE can be calculated for this fold (let's call this sMAPE_fold1).

The same process can then be repeated for all the folds. If there are 50 folds, then we will end up with 50 sMAPE values (sMAPE_fold1 to sMAPE_fold50).

The best hyperparameter selection can then be based on the average (or another statistic) value across these 50 values. 

In addition, it would be nice to have the raw values available to the user for all 50 folds and for each candidate hyperparameter. These can be used by the user to do further analysis if needed (and make any plots that they want), so the plotting can be outside the scope of `sktime`).

**Comment by ngupta23** [comment]
Also to clarify further, what I am proposing is similar to what is there in `scikit-learn` [`GridSearchCV`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)

Here is the excerpt from the scikit documentation which shows the mean test scores across the folds along with the standard deviation.

```python
{
'param_kernel': masked_array(data = ['poly', 'poly', 'rbf', 'rbf'],
                             mask = [False False False False]...)
'param_gamma': masked_array(data = [-- -- 0.1 0.2],
                            mask = [ True  True False False]...),
'param_degree': masked_array(data = [2.0 3.0 -- --],
                             mask = [False False  True  True]...),
'split0_test_score'  : [0.80, 0.70, 0.80, 0.93],
'split1_test_score'  : [0.82, 0.50, 0.70, 0.78],
'mean_test_score'    : [0.81, 0.60, 0.75, 0.85],
'std_test_score'     : [0.01, 0.10, 0.05, 0.08],
'rank_test_score'    : [2, 4, 3, 1],
'split0_train_score' : [0.80, 0.92, 0.70, 0.93],
'split1_train_score' : [0.82, 0.55, 0.70, 0.87],
'mean_train_score'   : [0.81, 0.74, 0.70, 0.90],
'std_train_score'    : [0.01, 0.19, 0.00, 0.03],
'mean_fit_time'      : [0.73, 0.63, 0.43, 0.49],
'std_fit_time'       : [0.01, 0.02, 0.01, 0.01],
'mean_score_time'    : [0.01, 0.06, 0.04, 0.04],
'std_score_time'     : [0.00, 0.00, 0.00, 0.01],
'params'             : [{'kernel': 'poly', 'degree': 2}, ...],
}
```

**Comment by fkiraly** [comment]
Ok, @ngupta23, so you are simply proposing that it should be possible to tune forecasters on a re-sample set-up where on each validation fold (out of many train/validation folds) you compute more than one ahead forecast?

The parameters for the tuned forecaster then would be selected by a double average, i.e., over folds and individual points in the validation splits?

If yes, that makes sense to me, and it surprises me that that wasn't already possible (though I have not extensivly tested the class).

PS: due to the temporal auto-correlation, the confidence intervals that you plot would be grossly misleading, so I would consider not outputting standard errors.

**Comment by ngupta23** [comment]
> The parameters for the tuned forecaster then would be selected by a double average, i.e., over folds and individual points in the validation splits?

Yes, sort of. I would like to generalize it though and say that the metric (whatever it is - would be some sort of average in most cases) can be calculated on each fold. Then the average of all these fold metrics can be used to select the best hyperparameters of the model.

> If yes, that makes sense to me, and it surprises me that that wasn't already possible (though I have not extensivly tested the class).

I got this error when I tried this in the current version.

https://github.com/alan-turing-institute/sktime/blob/79ac80ab65adb30fdf233e98433646c1989da2fc/sktime/forecasting/model_selection/_tune.py#L37

> PS: due to the temporal auto-correlation, the confidence intervals that you plot would be grossly misleading, so I would consider not outputting standard errors.

Yes, my bad on the CI. But the mean and std. dev. value over the folds would be quite useful.

**Comment by mloning** [comment]
This now works: 
```python
from sktime.datasets import load_airline
from sklearn.ensemble import RandomForestRegressor
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.model_selection import temporal_train_test_split, SlidingWindowSplitter, ForecastingGridSearchCV

import numpy as np
import pandas as pd

y = load_airline()

fh = [1, 2, 3]
y_train, y_test = temporal_train_test_split(y, fh=fh)
print(y.shape, y_train.shape[0], y_test.shape[0])

cv = SlidingWindowSplitter(initial_window=int(len(y_train) * 0.5), window_length=15)

regressor = RandomForestRegressor()
forecaster = make_reduction(regressor)

param_grid = {"estimator__n_estimators": [100, 200]}
gscv = ForecastingGridSearchCV(forecaster, cv=cv, param_grid=param_grid, verbose=1)
gscv.fit(y_train)
```
